### PR TITLE
[Hardware][Intel-Gaudi] Enable LoRA support for Intel Gaudi (HPU)

### PR DESCRIPTION
**Pull request by scsudhak-intel**
This PR enables support for LoRA on Intel Gaudi by adding HPU specific kernels in [`PunicaWrapperHPU`](https://github.com/vllm-project/vllm/blob/d51d66c3252107d5b986d2eab7af1c210dceb708/vllm/lora/punica_wrapper/punica_hpu.py#L11C7-L11C23) to handle LoRA computations more efficiently on the hardware.

### CODE REVIEW

**Comment by jeejeelee** [comment]
Can we do it like `CustomOp`, where the punicawrapper remains unchanged, and we decide which ops to call based on the hardware? I think if we do it this way, it might have better extensibility.

**Comment by SanjuCSudhakaran** [comment]
`GaudiPunicaWrapper` is a class inherited from `PunicaWrapper` which does the following.
1. Overrides the functions `add_lora`, `add_lora_packed_nslice` and `add_lora_logits`.
2. Adds `add_lora_embedding` instead of `add_expand` to handle LoRA B embedding computation.

I'd to get a little more clarity on the `CustomOp` s.

I see that classes like `RotaryEmbedding` using `CustomOp` to automatically handle the forward pass based on the backend device. Here `RotaryEmbedding` is inherited from `torch.nn.Module` and `CustomOp` handles which forward pass to call.

But here neither `PunicaWrapper` nor the ops in it are `torch.nn.modules`. So to use `CustomOp` the ops in `PunicaWrapper` would have to be rewritten to `torch.nn.Module`s which requires a lot of refactoring.

@jeejeelee Please correct if I am missing something here.

**Comment by jeejeelee** [comment]
@SanjuCSudhakaran  Thanks for your feedback. 
I'm mainly considering that other hardware may need to support LoRA in the future. If we support it in your current way,  it might lead to a lot of redundant conditionals. IMHO,  the main difference between different hardware should be in the implementation of the 6 LoRA ops. 
BTW, I actually have ideas and plans for extending LoRA support to other hardware. If you're interested, we can discuss this on Slack.

**Comment by SanjuCSudhakaran** [comment]
@jeejeelee Can you please confirm whether this is the slack you mentioned https://communityinviter.com/apps/vllm-dev/join-vllm-developers-slack

File: vllm/lora/layers.py [hunk 1]
```diff
@@ -1068,6 +1069,8 @@ def _get_logits(
         ).index_select(0, indices_padded).nan_to_num_(nan=float("-inf"),
                                                       posinf=float("inf"),
                                                       neginf=float("-inf")))
+        if current_platform.is_hpu():
+            lora_logits = lora_logits[:logits.shape[0], :]
```
Comment (jeejeelee) [inline vllm/lora/layers.py:1073]: As discussed on Slack,we should add comments about this

File: vllm/lora/punica_wrapper/punica_hpu.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        self,
+        y: Union[Tuple[torch.Tensor, ...], torch.Tensor],
+        x: torch.Tensor,
+        lora_a_stacked: Tuple[torch.Tensor, ...],
+        scale: float,
+        **kwargs,
+    ) -> None:
+        raise NotImplementedError
+
+    def add_expand(
```
Comment (jeejeelee) [inline vllm/lora/punica_wrapper/punica_hpu.py:76]: If we don't support `bias_enabled` and  fully_sharded_loras in this PR, we should  add related handling for them.  I think  [here](https://github.com/vllm-project/vllm/blob/main/vllm/worker/hpu_model_runner.py#L616)  is a good choice.

**Review by jeejeelee — COMMENTED** [review]
Some minor changes. Thank you very much for your contribution and patience.

File: vllm/lora/layers.py (hunk 1 above)
Comment (SanjuCSudhakaran) [reply vllm/lora/layers.py:1075]: Done

File: vllm/lora/punica_wrapper/punica_hpu.py (hunk 2 above)
Comment (SanjuCSudhakaran) [reply vllm/lora/punica_wrapper/punica_hpu.py:76]: Done

**Review by jeejeelee — APPROVED** [review]
Thank you very much for your contribution, and I especially appreciate your willingness to discuss the refactoring of the punicawrapper together

**Comment by jeejeelee** [comment]
@SanjuCSudhakaran  could you plz sync `main`?

**Pull request merged**